First draft for the thruster MCU

jhub04 opened the pull request.

jhub04 committed:
Generate peripheral drivers with MCC

jhub04 committed:
feat: enable LED as output

jhub04 committed:
feat/change pwm frequency to 50Hz

jhub04 committed:
feat/configure CAN1 TX and RX on PB12 and PB13 respectively

jhub04 committed:
feat: change CAN FD bit rates

jhub04 committed:
feat: configure SERCOM1 as I2C master

jhub04 committed:
feat: reconfigure tcc periods and prescalers

jhub04 committed:
feat: create table of thrusters' tcc channels

jhub04 committed:
feat: add method to set PWM dutycycle

jhub04 committed:
chore: disable treat warnings as errors

jhub04 committed:
feat: add methods to start and stop all thrusters

jhub04 committed:
feat: configure wdt

jhub04 committed:
feat: clear wdt after setting pwms

jhub04 committed:
feat: use can0 instead of can1

jhub04 committed:
feat: create can rx and tx callbacks for debugging purposes

jhub04 committed:
docs: add TODO reminder to remove CAN callback methods after CAN verification

jhub04 committed:
feat: change CAN-FD element size from 8 to 64 bytes

jhub04 committed:
feat: read can messages from Rx FIFO0

jhub04 committed:
feat: enable WDT

jhub04 committed:
feat: configure power manager

jhub04 committed:
feat: enter idle mode in main loop to reduce power consumption

jhub04 committed:
feat: add CAN message handler

jhub04 committed:
feat: add message handler function declaration and refactoring of consts

jhub04 committed:
refactor: make messages_to_read const

jhub04 committed:
refactor: separate app logic from main.c

jhub04 committed:
feat: make rx and tx callbacks static

jhub04 committed:
refactor: rename variable name for thruster pwm period

jhub04 committed:
refactor: use NVIC_SystemReset instead of manual wdt reset

jhub04 committed:
refactor: rename tcc_num to instance

jhub04 committed:
feat: set light pwm period to 20000ms

jhub04 committed:
feat: create method to set light pwm

jhub04 committed:
feat: add state set_light_pwm

jhub04 committed:
feat: pet watchdog after setting light pwm

jhub04 committed:
refactor: use switch statement instead of hardcoded tcc instance for lights

jhub04 committed:
refactor: rename light to lights

jhub04 committed:
docs: add TODO reminder to clamp tcc value

jhub04 committed:
style: remove leading whitespace

jhub04 committed:
feat: add clamping for thrusters and lights

jhub04 committed:
refactor: extract switch for setting tcc into function

jhub04 committed:
refactor: rename duty_cycle to us and tcc_value to ticks

jhub04 committed:
refactor: rename tcc period to tcc period_ticks

jhub04 committed:
refactor: extract separate function to convert pulse us to ticks

jhub04 committed:
refactor: rename pwm period consts to us for consistent naming

jhub04 committed:
refactor: rename us to pulse_us

jhub04 committed:
fix: stop_thrusters neutralizes thrusters

The function now neutralizes the thrusters by setting each of them to
1500us instead of just shutting down the tcc instances. This will also
fix the issue of that the camera lights also would turn off when
stop_thrusters is called.

jhub04 committed:
style: use u instead of mu symbol in comments

jhub04 committed:
feat: set start_thrusters to no-op

jhub04 committed:
feat: add method to turn lights off

jhub04 committed:
feat: add turn lights off event

jhub04 committed:
feat: configure ADC0 with sleep during standby

jhub04 committed:
feat: set up adc variables

jhub04 committed:
feat: enable ADC0 in app_init

jhub04 committed:
feat: create method to read thruster's current draw

jhub04 committed:
feat: break out of function if overcurrent detected

jhub04 committed:
feat: read current draw after each CAN receive

I don't know whether this is the optimal approach or not. I will do some
more research. Interrupts might be better if the operation is time
effective.

jhub04 committed:
refactor: add function prototype and move read_current_draw

jhub04 committed:
fix: adc actually converts now

jhub04 committed:
feat: configure adc0 for adc dma sequencing

jhub04 committed:
feat: enable result ready input for adc

jhub04 committed:
feat: configure tc0 witt timer period overflow event enabled

jhub04 committed:
feat: configure dmac channels 0 and 1

jhub04 committed:
Enable adc input control

jhub04 committed:
feat: configure evsys user and event channel

jhub04 committed:
feat: enable adc start event input on rising edge

jhub04 committed:
feat: disable run during standby

jhub04 committed:
feat: configure adc dma sequencing

jhub04 committed:
feat: implement minimal overcurrent check

jhub04 committed:
feat: reduce callback time complexity

jhub04 committed:
refactor: rename variable i to sample

jhub04 committed:
style: use consistent curly bracket indentation

jhub04 committed:
refactor: make rated current unsigned

jhub04 committed:
feat: add logging

jhub04 committed:
fix: valid printf

jhub04 committed:
fix: change adc_seq_regs to uint32_t array

jhub04 committed:
feat: add adc average control

jhub04 committed:
feat: enable adc sample averaging

jhub04 committed:
chore: remove adc averaging to avoid merge conflicts in the development process

jhub04 committed:
refactor: make clamp function easier to read

jhub04 committed:
refactor: remove unused global variable input_voltage

jhub04 committed:
feat: add realistic values for current monitoring

jhub04 committed:
feat: configure adc pins for testing

jhub04 committed:
feat: add debugging prints

jhub04 committed:
test: current readings on a simple circuit

jhub04 committed:
feat: configure CAN1

jhub04 committed:
Reconfigure can states and import some function from demo

jhub04 committed:
fix: handle can frames correctly

jhub04 committed:
feat: implement can id check

jhub04 committed:
feat: improve can receive callback

jhub04 committed:
feat: add debugging prints to can transmit callback

jhub04 committed:
refactor: remove unused demo functions

jhub04 committed:
style: remove outdated comments

jhub04 committed:
refactor: remove unused functions turn_thrusters_on and turn_lights_on

jhub04 committed:
feat: configure tcc0 pin

jhub04 committed:
feat: update can1 configuration

jhub04 committed:
feat: change tcc0/1 period to 75000ms

jhub04 committed:
fix: initialize PWM

jhub04 committed:
feat: only enter message_handler after received flag is true

jhub04 committed:
refactor: add debugging statements

jhub04 committed:
Set thrusters and lights to neutral on initialization

jhub04 committed:
refactor: check adc_dma_done before entering overcurrent check

jhub04 committed:
Merge branch 'main' into feat/thruster-mcu

jhub04 committed:
refactor: use plib function to retrieve pwm period instead of hardcoded values

jhub04 committed:
fix: mcc merge conflict

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+    if (overcurrent_fault) {
+        overcurrent_fault = false;
+        turn_thrusters_off();
+    }
```
Why is turn_thrusters_off not in the for loop? Would make sense to turn thrusters off as soon as we detect overcurrent

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+
+static inline uint16_t clamp(uint16_t value, uint16_t low, uint16_t high) {
+    if (value < low) {
+        return low;
+    } else if (value > high) {
+        return high;
+    } else {
+        return value;
+    }
+    
+}
+
+static inline void tcc_write(uint8_t instance, uint8_t channel, uint32_t ticks) {
+    switch (instance) {
+        case 0: TCC0_PWM24bitDutySet(channel, ticks); break;
+        case 1: TCC1_PWM24bitDutySet(channel, ticks); break;
+        case 2: TCC2_PWM16bitDutySet(channel, (uint16_t)ticks); break; // Not used with current mapping
+        default: break;
+    }
```
```suggestion
    switch (instance) {
        case 0:
            TCC0_PWM24bitDutySet(channel, ticks);
            break;
        case 1:
            TCC1_PWM24bitDutySet(channel, ticks);
            break;
        case 2:
            TCC2_PWM16bitDutySet(channel, (uint16_t)ticks);
            break;  // Not used with current mapping
        default:
            break;
    }

```

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+
+    switch (event) {
+        case TURN_THRUSTERS_OFF:
+            printf("Command: TURN_THRUSTERS_OFF\r\n");
```
Lots of print statements in this code. This is fine for debugging, but printf is actually a really slow function. Once you have tested CAN i suggest you remove all these printfs

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+    if (id < CAN_EVENT_ID_BASE || id >= (CAN_EVENT_ID_BASE + 5)) {
+        printf("Unknown CAN ID, ignoring\r\n");
+        return;
+    }
```
Look into CAN ID filtering in MCC. There should be an option to only accept IDs we are interested in. We do not need this if check then

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+    }
+    
+    uint8_t event = (uint8_t)(id - CAN_EVENT_ID_BASE);
+    const uint8_t *pData = rxBuf->data;
```
Consider having the CAN events enum map directly to the IDs instead of having to subtract like this. The reason I did it on the previous gripper and thruster was to both be compatible with I2C and CAN at the same time.

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+    uint8_t instance;
+    uint8_t channel;
+    uint32_t period_ticks;
+} Light;
```
Structs in C use snake case. Also in this case I would recommend not using typedef. You then have to declare a instance of the struct like this: Makes it a lot clearer that we are referring to a struct

```c
// Instead of
Thruster thrusters[8];
// Do
struct thruster thrusters[8];
```

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+    APP_STATE_CAN_XFER_SUCCESSFUL,
+    APP_STATE_CAN_XFER_ERROR,
+    APP_STATE_CAN_USER_INPUT
+} APP_CAN_STATES;
```
Where are these states used? Do not think we need this enum

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+static volatile bool can_message_received = false;
+
+/* Variable to save CAN application state*/
+volatile static APP_CAN_STATES state = APP_STATE_CAN_USER_INPUT;
```
Same as the message above, seems like this variable is never used

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+
+        case RESET:
+            printf("Command: RESET - Resetting system...\r\n");
+            /* Force a watchdog reset */
```
```suggestion
            /* Force a system reset */
```

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+
+static inline uint16_t clamp(uint16_t value, uint16_t low, uint16_t high);
+static inline void tcc_write(uint8_t instance, uint8_t channel, uint32_t ticks);
+static inline uint32_t us_to_ticks(uint32_t period_ticks, uint16_t us, uint32_t frame_us);
```
These functions should have doxygen comments. See the gripper for reference

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+
+/* --- Private function prototypes --- */
+static void set_thruster_pwm(const uint8_t *data);
+static void set_light_pwm(const uint8_t *data);
```
These functions should have doxygen comments. Se gripper for reference

forrisdahl commented on thruster_mcu/src/main.c:
```diff
@@ -0,0 +1,28 @@
+        App_Task();
+        
+        /* Maintain state machines of all polled MPLAB Harmony modules. */
+        SYS_Tasks ( );
```
You can remove this function call from main

forrisdahl commented on thruster_mcu/src/app/app.h:
```diff
@@ -0,0 +1,23 @@
+void App_Init(void);
+
+/* Handle incoming frames and actions */
+void App_Task(void);
```
Use snake case for these functions

forrisdahl reviewed (changes requested):
Nice work! 

Left some comments and suggestions

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+/* --- Constants --- */
+static const uint32_t TCC0_PERIOD               = TCC0_PWM24bitPeriodGet();
+static const uint32_t TCC1_PERIOD               = TCC1_PWM24bitPeriodGet();
+static const uint32_t TCC2_PERIOD               = TCC2_PWM16bitPeriodGet();
```
I am slightly concerned about calling these functions potentially before TCCx is initialized. Might be completely safe, but I think I would either make sure we call these function after the init is done or hardcoding the value

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,335 @@
+static const float    ADC_VREF                  = 3.3f;
+static const float    G_IMON                    = 18.18e-6f; // Amplifier gain 18.18 uA/A -> in A/A
+static const float    R_IMON                    = 2550.0f;   // 2.55 kilo ohms resistor
+static const uint8_t  THRUSTER_RATED_CURRENT    = 15U;    // From TSD7 datasheet  
```
Whenever possible, try to limit the scope of constants. Move these constants to the check overcurrent function

forrisdahl reviewed (changes requested):
Left some more suggestions while I remember them.

jhub04 committed:
refactor: call turn_thrusters_off directly instead of setting flag

jhub04 committed:
style: space out switch statement

jhub04 committed:
refactor: remove all printf's

jhub04 committed:
revert: use hardcoded values for tccx period

jhub04 committed:
refactor: remove unused can state enum

jhub04 committed:
feat: configure can to automatically handle id filtering

jhub04 committed:
refactor: directly map can states enum to the ID's

jhub04 committed:
refactor: use snake-case for structs and use structs instead of typedefs

jhub04 committed:
style: more precise comment

jhub04 committed:
refactor: remove sys_tasks function in main loop

jhub04 committed:
docs: add Doxygen comments to utility functions

jhub04 committed:
refactor: rename functions to snake_case for consistency

jhub04 committed:
refactor: move hardware constants to local scope in check_overcurrent()

jhub04 commented:
All the issues should now be resolved. I don't know if the Doxygen comment for message_handler() is necessary, I added it because I don't think the function name is self documenting. I can't really think of another name that is either. Also should I transmit a CAN frame if overcurrent is detected?

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,360 @@
+    uint8_t instance;
+    uint8_t channel;
+    uint32_t period_ticks;
+};
```
These structs are identical. I suggest removing one of them and instead having a generic name

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,360 @@
+    
+    // Pet the watchdog after applying updates
+    WDT_Clear();
+}
```
These functions are almost identical. I suggest having only one function where the thruster table (or light table) is passed as an input to the function and the how many tcc instances we are going to set.

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,360 @@
+static uint8_t Can1MessageRAM[CAN1_MESSAGE_RAM_CONFIG_SIZE] __attribute__((aligned(32)));
+
+#define WRITE_ID(id) (id << 18)
+#define READ_ID(id) (id >> 18)
```
defines should be after include and before const declarations

forrisdahl commented:
> All the issues should now be resolved. I don't know if the Doxygen comment for message_handler() is necessary, I added it because I don't think the function name is self documenting. I can't really think of another name that is either. Also should I transmit a CAN frame if overcurrent is detected?

Yes, I think we should transmit a CAN frame if overcurrent is detected. This could be logged by some program running on Orin or the Raspberry pi

jhub04 committed:
refactor: move defines after include and before constants

jhub04 committed:
refactor: combine thruster and light struct into one generic struct

jhub04 committed:
refactor: combine set_thruster_pwm and set_light_pwm into one generic function

jhub04 committed:
refactor: combine turn_thrusters_off and turn_lights_off into one generic function

jhub04 committed:
docs: add TODO reminder to send can transmit on overcurrent fault

jhub04 committed:
feat: transmits fault message over CAN if overcurrent flagged

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,363 @@
+    
+    memset(txFiFo, 0x00, CAN1_TX_FIFO_BUFFER_SIZE);
+    txBuffer->id = WRITE_ID(0x45A); // Just a random ID
+    txBuffer->dlc = 0x7U;           // DLC 7 -> 7 Byte Payload
```
I think since the payload is below 8 DLC will equal the byte payload. However for larger payloads there is a table that says which DLC is value is used for what length

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,363 @@
+    
+    txBuffer->data[0] = thruster_id; 
+    memcpy(&txBuffer->data[1], &current, sizeof(float)); 
+    memcpy(&txBuffer->data[5], &adc_raw, sizeof(uint16_t));
```
Be careful here, you are currently sending the data in packed format. This reduces the total size but makes it harder to parse later since the data is unalligned. This is something i am currently dealing a lot with on the MRU and something we should avoid here. Add padding to make sure that the data is alligned correctly

jhub04 replied:
Okay I will start by just sending 64 bytes with a lot of excessive padding, and then I can cut down on that when I get it working.

forrisdahl replied:
The important padding here is the padding *between* the different elements

* `uint16_t` must be **2-byte aligned**
* `float` must be **4-byte aligned**

This affects how the compiler inserts padding to ensure proper alignment.

For more detail on structure padding and alignment, see: https://www.geeksforgeeks.org/c/structure-member-alignment-padding-and-data-packing/

jhub04 replied:
Should be done now:)

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,363 @@
+}
+
+
+static void set_pwm_outputs(const uint8_t *data, struct pwm_output *outputs, size_t count, uint16_t min_us, uint16_t max_us, uint32_t frame_us) {
```
When we are moving to SAMC21 TCC2 might have different min_us and max_us than TCC0 and TCC1. Consider adding an additional field in the struct containing the min and max value since these are constants defined at compile time

jhub04 replied:
Just to clarify, are you anticipating that TCC2 will have different hardware imposed min/max pulse widths than TCC0/TCC1 on the SAMC21? Or is this just for general flexibility?

forrisdahl replied:
No, but I think currently this function has more inputs that we need. We know the min,max CC values on compile time so it would make sense to include it as a part of the thruster table.

I think when I commented it I thought that min_us was the tcc value and not the time in us.

forrisdahl commented on thruster_mcu/src/app/app.c:
```diff
@@ -0,0 +1,363 @@
+
+static void set_pwm_outputs(const uint8_t *data, struct pwm_output *outputs, size_t count, uint16_t min_us, uint16_t max_us, uint32_t frame_us) {
+    for (size_t i = 0; i < count; i++) {
+        uint16_t pulse_us = ((uint16_t)data[2U * i] << 8) | (uint16_t)data[2U * i + 1U];
```
I also think we can simplify this. We are going to send data over CAN in little endian order. This means that you can simply cast the array to uint16_t array instead of manually having to rebuild the array

forrisdahl reviewed (changes requested):
Really solid progress! Here is some more suggestions

jhub04 committed:
fix: initialize txBuffer to txFiFo after zeroing out

jhub04 committed:
feat: transmit 64 bytes instead of 7 bytes

jhub04 committed:
refactor: log error if can message transmit fails

jhub04 committed:
feat: enable can timestamps

jhub04 committed:
fix: read CAN data in little-endian format instead of big-endian

jhub04 committed:
refactor: send 8 byte payload instead of 64 bytes

jhub04 committed:
refactor: include min_us, max_us, neutral_us and frame_us as part of pwm_output struct

jhub04 committed:
feat!: migrate from SAME54P20A to SAMC21J18A

BREAKING CHANGE: MCU platform changed from SAME54 to SAMC21

SAME54 3.3V output is insufficient for thruster ESC control. SAMC21
provided sufficient 5V voltage.

jhub04 committed:
fix: reenable WDT

jhub04 committed:
feat: configure DMAC for adc sleepwalking

jhub04 committed:
refactor: initialize adc before dma

jhub04 committed:
feat: add simple adc dma callback

jhub04 committed:
feat: register DMAC callback and transfer protocol

jhub04 committed:
feat: adc averaging of 1024 samples

jhub04 committed:
feat: configure pins for the SAMC21 Xplained Pro

jhub04 committed:
feat: enable starte event input on rising edge

jhub04 committed:
feat: configure EIC

jhub04 committed:
feat: configure eic interrupt line on debboard pin

jhub04 committed:
feat: configure ISR for EXTINT0

jhub04 committed:
feat: configure 8 EXTINT channels

jhub04 committed:
feat: register callback for all EXTINT channels

jhub04 committed:
fix: correct function call

jhub04 committed:
fix: re-arm DMA for next conversion

jhub04 committed:
refactor: uncomment check_overcurrent()

jhub04 committed:
fix: miscellaneous issues

jhub04 committed:
fix: reconfigure dmac

jhub04 committed:
fix: create new project

jhub04 committed:
fix: duplicated project

jhub04 committed:
fix: add app.h include in main.c

jhub04 committed:
refactor: remove comment to remember average sampling

jhub04 committed:
fix: implement correct padding for thruster fault transmission

jhub04 committed:
feat: configure TCC2

jhub04 committed:
feat: configure eight independent PWM output pins

jhub04 committed:
feat: configure killswitch pin as input

jhub04 committed:
feat: configure TC3 and light PWM pi (PWM_9)

jhub04 committed:
feat: enable external interrupt for killswitch

jhub04 committed:
feat: configure PG_THx pins as input

jhub04 committed:
feat: configure CAN1 TX and RX pins

jhub04 committed:
feat: configure FLT_THx pins

jhub04 committed:
feat: configure IMON_THx

jhub04 committed:
feat: configure RESET_TH

jhub04 committed:
feat: add PWM writer for TCC2

jhub04 committed:
fix: update pwm_ouput struct with updated pwm outputs

jhub04 committed:
feat: enable ADC AIN pins

jhub04 committed:
feat: configure TC3 with MPWM

jhub04 committed:
fix: correct period_ticks for pwm_outputs

jhub04 committed:
fix: separate TCC and TC outputs

jhub04 committed:
fix: only update CC1 for TC

jhub04 committed:
fix: correct instance and channel of lights struct

jhub04 committed:
feat: adjust TC3 period

jhub04 committed:
fix: initialize TCC2

jhub04 committed:
fix: initialize TC3

jhub04 committed:
test: simple pmw signal for TCC0 and TCC1

jhub04 committed:
fix: clear wdt after pwm set

jhub04 committed:
test: current attempt of TC3 MPWM generation

jhub04 committed:
refactor: update pins according to new schematic

jhub04 committed:
feat: reconfigure pin config to utilize all interrupts

The PCB schematic has been redone to account for interrupts lines as
well such that all the PG and FLT pins have distinct EXTINT numbers, KS
is using NMI

jhub04 committed:
feat: enable NMI control for EIC

jhub04 committed:
feat: enable callbacks for killswitch and flt pins

jhub04 committed:
feat: remove software overcurrent check and log IMON output

jhub04 committed:
feat: simplify send_thruster_fault() to just log thruster no

jhub04 committed:
fix: correct implementation of raw ADC data conversion

jhub04 committed:
fix: use WO1 as output channel for TC3

jhub04 committed:
docs: update TCC testing reminder

jhub04 committed:
fix: configure usart and reset thrusters pins

jhub04 committed:
fix: correct configuration of TCC0

jhub04 committed:
fix: correct configuration of all TCC instances

jhub04 committed:
fix: correct CAN config

jhub04 committed:
fix: update R_IMON to correct value

jhub04 committed:
fix: configure ADC sleepwalking correctly

jhub04 committed:
refactor: move ADC conversion start into app_task()

jhub04 committed:
refactor: move initial conversion start into app_init()

jhub04 committed:
fix: disable adc result averaging

jhub04 committed:
feat: enable interrupts for pgood pins

jhub04 committed:
feat: log_current() prints correct IMON order

jhub04 committed:
feat: make thruster and light table const

jhub04 committed:
fix: update efuse resistance to updated value

jhub04 committed:
feat: add corresponding pwm value for adc readings

jhub04 committed:
feat: update thruster table to go in increasing thruster order

jhub04 committed:
docs: remove unneccessary comment

jhub04 committed:
feat: dispatch hardware event cna messages

jhub04 committed:
refactor: delete dead send_thruster_fault function

jhub04 committed:
feat: send can message with current measurements

jhub04 committed:
refactor: move can tx logic out of log_current()

jhub04 committed:
feat: handle lights in set_pwm_neutral()

jhub04 committed:
fix: matching channel numbers for thruster id's

jhub04 committed:
feat: increase WDT period

jhub04 committed:
refactor: remove pure test functions

jhub04 committed:
refactor: remove newline print
